mqtt.go

```diff
@@ -19,598 +24,613 @@ const (
 
 var mqttDialer = &net.Dialer{Timeout: mqttConnectTimeout}
 
-type (
-	mqttMsgHandler func(*packet.PublishPacket) error
-
-	mqttSubscription struct {
-		topic      string
-		msgHandler mqttMsgHandler
-	}
-
-	mqttConn struct {
-		conn           net.Conn
-		stream         *packet.Stream
-		dc             *DeviceContext
-		packetID       uint16
-		subs           map[string]mqttSubscription
-		command        chan<- *DeviceCommand
-		event          <-chan *DeviceEvent
-		bulkData       <-chan *DeviceBulkData
-		syncedBulkData <-chan *DeviceSyncedBulkData
-		kvSync         chan<- *keyValueSync
-		kvPush         <-chan *keyValueSync
-		err            chan error
-		doPing         chan time.Duration
-		outPacket      chan packet.Packet
-		puback         chan *packet.PubackPacket
-		pingresp       chan *packet.PingrespPacket
-		stopPublisher  chan bool
-		wgWrite        sync.WaitGroup
-		wgRead         sync.WaitGroup
-	}
-)
+// QoS level of message delivery. This is used in sending events to MODE.
+type QOSLevel int
 
-func (mc *mqttConn) close() {
-	close(mc.stopPublisher) // tell publisher to quit
-	close(mc.doPing)        // tell pinger to quit
-
-	mc.wgWrite.Wait() // wait for event processor and pinger to finish
-
-	// Attempt graceful disconnect.
-	mc.outPacket <- packet.NewDisconnectPacket()
-	close(mc.outPacket)
+const (
+	// QoS 0 - message delivery is not guaranteed.
+	QOSAtMostOnce QOSLevel = iota
 
-	mc.wgRead.Wait() // wait for packet reader to finish
-}
+	// QoS 1 - message is delivered at least once, but duplicates may happen.
+	QOSAtLeastOnce
 
-func (mc *mqttConn) getErrorChan() chan error {
-	return mc.err
-}
+	// QoS 2 - message is always delivered exactly once. This is currently not supported.
+	QOSExactlyOnce
+)
 
-func (mc *mqttConn) ping(timeout time.Duration) {
-	mc.doPing <- timeout
-}
+type (
+	// Data to send in the channel to the client when we receive data
+	// published for our subscription
+	MqttSubData struct {
+		topic string
+		data  []byte
+	}
+
+	// Data to send in the channel to the client when we receive the
+	// PUBACK on publish, or an error
+	MqttQueueResult struct {
+		PacketId uint16
+		Err      error
+	}
+
+	// For users of MqttClient. Three responsibilities (Oops. single
+	// responsibility, blah blah.):
+	// - data needed to establish a connection
+	// - Callback method to create data to publish
+	// - channels to communicate data back to a receiver.
+	// These methods should be const, since we might call them frequently.
+	// I think this could be broken up - maybe a struct for the channels, with
+	// another interface (and the struct implement the interface if desired).
+	// But this is sufficient for now.
+	MqttDelegate interface {
+		// Returns authentication information
+		AuthInfo() (username string, password string)
```

Comment:
Since we use the device ID in the topic strings, I put user and password into the delegate. But, that is our implementation detail. We can just pass the parameters in Connect().

Reply:
I understand. Thank you.

Reply:
I understand. Thank you.